genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/FileSystemFetchingCacheService.java

```diff
@@ -0,0 +1,343 @@
+ * <p>
+ * Deletion of older versions
+ *
+ * @author standon
+ * @since 4.0.0
+ */
+@Service
+@Lazy
+@Slf4j
+@Component
+class FileSystemFetchingCacheService implements FetchingCacheService {
+
+    public static final String LOCK_FILE_NAME = "lock";
+    public static final String DATA_FILE_NAME = "data";
+    public static final String DATA_DOWNLOAD_FILE_NAME = "data.tmp";
+    protected final ResourceLoader resourceLoader;
+    protected final File cacheDirectory;
+    protected final LockingFactory lockingFactory;
+    private final ExecutorService oldResourcesCleanUpService = Executors.newFixedThreadPool(1);
+
+
+    FileSystemFetchingCacheService(
+        final ResourceLoader resourceLoader,
+        final ArgumentDelegates.CacheArguments cacheArguments,
+        final LockingFactory lockingFactory
+    ) {
+        this.resourceLoader = resourceLoader;
+
+        this.cacheDirectory = cacheArguments.getCacheDirectory();
+
+        this.lockingFactory = lockingFactory;
+
+        createDirectoryStructureIfNotExists(cacheDirectory);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void get(final URI sourceFileUri, final File destinationFile) throws DownloadException, IOException {
+        lookupOrDownload(sourceFileUri, destinationFile);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void get(final Set<Pair<URI, File>> sourceDestinationPairs) throws DownloadException, IOException {
+        for (final Pair<URI, File> sourceDestinationPair : sourceDestinationPairs) {
+            get(sourceDestinationPair.getKey(), sourceDestinationPair.getValue());
+        }
+    }
+
+    protected void lookupOrDownload(
+        final URI sourceFileUri,
+        final File destinationFile
+    ) throws DownloadException, IOException {
+
+        final String uriString = sourceFileUri.toASCIIString();
+
+        log.info("Lookup: {}", uriString);
+
+        // Unique id to store the resource on local disk
+        final String resourceCacheId = getResourceCacheId(sourceFileUri);
+
+        // Get a handle to the resource
+        final Resource resource = resourceLoader.getResource(uriString);
+
+        if (!resource.exists()) {
+            throw new DownloadException("Resource not found: " + uriString);
+        }
+
+        final long resourceLastModified = resource.lastModified();
+
+        //Create the resource version dir in cache if it does not exist
+        final File cacheResourceVersionDir = getCacheResourceVersionDir(
+            resourceCacheId,
+            resourceLastModified
+        );
+
+        createDirectoryStructureIfNotExists(cacheResourceVersionDir);
+
+        //Handle to the resource cached locally
+        final File cachedResourceVersionDataFile = getCacheResourceVersionDataFile(
+            resourceCacheId,
+            resourceLastModified
+        );
+
+        try (
+            final FileChannel lockFileChannel = getLockFileChannel(
```

Comment:
Beating a dead horse but again pointing out here the very specific implementation here and how it effectively renders the `Lock` interface useless.